### PR TITLE
perf: stable row id prefilter

### PR DESCRIPTION
**Pull request by wjones127**
Cache the row id mask prefilter and make it faster to construct.

### CODE REVIEW

File: rust/lance-table/src/rowids.rs [hunk 1]
```diff
@@ -753,4 +761,60 @@ mod test {
...
+            },
+            U64Segment::Range(9..10),
+            U64Segment::RangeWithBitmap {
+                range: 10..15,
+                bitmap: [true, false, false, false, true].as_slice().into(),
+            },
+            U64Segment::Array(vec![35].into()),
+        ]);
+        assert_eq!(sequence, expected);
+    }
```
Comment (westonpace) [inline rust/lance-table/src/rowids.rs:819]: What happens if you mask out every row id in a segment?

File: rust/lance-table/src/rowids/segment.rs [hunk 2]
```diff
@@ -392,6 +392,62 @@ impl U64Segment {
...
+            Self::RangeWithBitmap { .. } => true,
+            Self::SortedArray(_) => true,
+            Self::Array(_) => false,
+        };
+        // To get minimum, need to find the first value that is not masked.
+        let first_unmasked = (0..self.len())
+            .zip(positions.iter().cycle())
+            .filter(|(sequential_i, i)| **i != *sequential_i as u32)
+            .map(|(sequential_i, _)| sequential_i)
+            .next()
```
Comment (westonpace) [inline rust/lance-table/src/rowids/segment.rs:417]: Can you use [`find`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.find) or `find_map`?

File: rust/lance-table/src/rowids/segment.rs [hunk 3]
```diff
@@ -392,6 +392,62 @@ impl U64Segment {
...
+        let max = self.get(last_unmasked).unwrap();
+
+        let stats = SegmentStats {
+            min,
+            max,
+            count,
+            sorted,
+        };
+
+        let make_new_iter = || {
```
Comment (westonpace) [inline rust/lance-table/src/rowids/segment.rs:437]: Why is a named closure needed here?  Can't you just execute it?

File: rust/lance/src/index/prefilter.rs [hunk 4]
```diff
@@ -119,49 +122,59 @@ impl DatasetPreFilter {
...
+            dataset: &Dataset,
+        ) -> Result<Vec<(Arc<RowIdSequence>, Option<Arc<DeletionVector>>)>> {
+            stream::iter(dataset.get_fragments())
+                .map(|frag| async move {
+                    let row_ids = load_row_id_sequence(dataset, frag.metadata());
+                    let deletion_vector = frag.get_deletion_vector();
+                    let (row_ids, deletion_vector) = join!(row_ids, deletion_vector);
+                    Ok::<_, crate::Error>((row_ids?, deletion_vector?))
+                })
+                .buffer_unordered(10)
```
Comment (westonpace) [inline rust/lance/src/index/prefilter.rs:141]: I see you  just moved this function but maybe let's get rid of the random `10`?

File: rust/lance/src/index/prefilter.rs [hunk 5]
```diff
@@ -119,49 +122,59 @@ impl DatasetPreFilter {
...
+        session
+            .file_metadata_cache
+            .get_or_insert(&path, move |_| {
+                let dataset = dataset.clone();
+                async move {
+                    let row_ids_and_deletions = load_row_ids_and_deletions(&dataset).await?;
+
+                    // The process of computing the final mask is CPU-bound, so we spawn it
+                    // on a blocking thread.
+                    let allow_list = tokio::task::spawn_blocking(move || {
```
Comment (westonpace) [inline rust/lance/src/index/prefilter.rs:155]: Perhaps we should use `spawn_cpu` here?

**Review by westonpace — APPROVED** [review]
Looks good, nice work on the `mask` function.

File: rust/lance-table/src/rowids/segment.rs (hunk 3 above)
Comment (wjones127) [reply rust/lance-table/src/rowids/segment.rs:437]: I could, but that just wouldn't be convoluted enough. :smile:

I'll simplify this.

File: rust/lance/src/index/prefilter.rs (hunk 4 above)
Comment (wjones127) [reply rust/lance/src/index/prefilter.rs:141]: What is our best practice right now. I don't love this, but I just don't know what our current preferred pattern is right now.

Comment (westonpace) [reply rust/lance/src/index/prefilter.rs:142]: Since it's mostly I/O driven I'd recommend `dataset.object_store().io_parallelism()`?

**Pull request merged**